src/main/kotlin/com/chattriggers/ctjs/minecraft/listeners/ClientListener.kt

```diff
@@ -193,73 +195,83 @@ object ClientListener : Initializer {
             )
 
             if (event.isCancelled()) ActionResult.FAIL else ActionResult.PASS
         }
 
-        AttackEntityCallback.EVENT.register { _, _, _, entity, _ ->
+        AttackEntityCallback.EVENT.register { player, _, hand, entity, _ ->
+            if (!player.world.isClient) return@register ActionResult.PASS
             val event = CancellableEvent()
 
             TriggerType.PLAYER_INTERACT.triggerAll(
                 PlayerInteraction.ATTACK_ENTITY,
                 Entity.fromMC(entity),
+                hand,
                 event,
             )
 
             if (event.isCancelled()) ActionResult.FAIL else ActionResult.PASS
         }
 
-        PlayerBlockBreakEvents.BEFORE.register { _, _, pos, state, _ ->
-            val event = CancellableEvent()
-
+        CTEvents.BREAK_BLOCK.register { pos, hand ->
             TriggerType.PLAYER_INTERACT.triggerAll(
                 PlayerInteraction.BREAK_BLOCK,
-                Block(BlockType(state.block), BlockPos(pos)),
-                event,
+                World.getBlockAt(BlockPos(pos)),
+                hand,
+                CancellableEvent(),
             )
-
-            !event.isCancelled()
         }
 
-        UseBlockCallback.EVENT.register { _, _, _, hitResult ->
+        UseBlockCallback.EVENT.register { player, _, hand, hitResult ->
+            // TODO: double firing
+            if (!player.world.isClient) return@register ActionResult.PASS
             val event = CancellableEvent()
 
             TriggerType.PLAYER_INTERACT.triggerAll(
                 PlayerInteraction.USE_BLOCK,
                 World.getBlockAt(BlockPos(hitResult.blockPos)).withFace(BlockFace.fromMC(hitResult.side)),
+                hand,
                 event,
             )
 
             if (event.isCancelled()) ActionResult.FAIL else ActionResult.PASS
         }
 
-        UseEntityCallback.EVENT.register { _, _, _, entity, _ ->
+        UseEntityCallback.EVENT.register { player, _, hand, entity, _ ->
+            if (!player.world.isClient) return@register ActionResult.PASS
             val event = CancellableEvent()
 
             TriggerType.PLAYER_INTERACT.triggerAll(
                 PlayerInteraction.USE_ENTITY,
                 Entity.fromMC(entity),
-                event
+                hand,
+                event,
             )
 
             if (event.isCancelled()) ActionResult.FAIL else ActionResult.PASS
         }
 
-        UseItemCallback.EVENT.register { _, _, _ ->
+        UseItemCallback.EVENT.register { player, _, hand ->
+            if (!player.world.isClient) return@register TypedActionResult.pass(null)
             val event = CancellableEvent()
 
+            val stack = player.getStackInHand(hand)
+
             TriggerType.PLAYER_INTERACT.triggerAll(
                 PlayerInteraction.USE_ITEM,
-                // TODO,
-                event
+                if (stack.isEmpty) null else Item(stack),
+                hand,
```

Comment:
Are any of the interaction events applicable to both hands? It doesn't seem like it but I could be wrong. That's why I originally had the field, so you could just check `type.isLeftHand` or whatever

Reply:
Yeah the UseBlock, UseEntity, and UseItem fire on both main and off hand.